### PR TITLE
use switchOnFirst replacing HeadAndTail

### LINKED ISSUE #509: investigate if switchOnFirst may replace HeadAndTail.
image

https://github.com/scalecube/scalecube-services/blob/a9761776973df8d990658eeffca7d5025f58b209/services/src/main/java/io/scalecube/services/ServiceCall.java#L204

https://github.com/reactor/reactor-core/pull/1423/files#diff-c590f0f2fbe1d419bdccc521a1894278R186

https://github.com/reactor/reactor-core/pull/1423/files#diff-1496e2df144760bd5a12dbd0c149f40eR148

### PR DESCRIPTION
**Pull request by ronenhamias**
#509
1. update version 3.2.5.RELEASE
2. refactor and use switchOnFirst.

### CODE REVIEW

**Comment by ronenhamias** [comment]
reactor.core.Exceptions$BubblingException: java.lang.IllegalStateException: FluxSwitchOnFirst allows only one Subscriber
	at reactor.core.Exceptions.bubble(Exceptions.java:154)
	at reactor.core.publisher.Operators.onErrorDropped(Operators.java:512)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120)
	at reactor.core.publisher.Operators.error(Operators.java:181)
	at reactor.core.publisher.FluxSwitchOnFirst$SwitchOnFirstInner.subscribe(FluxSwitchOnFirst.java:160)
	at reactor.core.publisher.Flux.subscribe(Flux.java:7743)
	at reactor.core.publisher.FluxSwitchOnFirst$SwitchOnFirstInner.onComplete(FluxSwitchOnFirst.java:268)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:144)
	at reactor.core.publisher.FluxJust$WeakScalarSubscription.request(FluxJust.java:101)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
	at reactor.core.publisher.FluxSwitchOnFirst$SwitchOnFirstInner.onSubscribe(FluxSwitchOnFirst.java:168)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90)
	at reactor.core.publisher.FluxJust.subscribe(FluxJust.java:70)
	at reactor.core.publisher.FluxMapFuseable.subscribe(FluxMapFuseable.java:63)
	at reactor.core.publisher.FluxSwitchOnFirst.subscribe(FluxSwitchOnFirst.java:63)
	at reactor.core.publisher.FluxMap.subscribe(FluxMap.java:62)
	at reactor.core.publisher.Flux.subscribe(Flux.java:7743)
	at reactor.test.DefaultStepVerifierBuilder$DefaultStepVerifier.verify(DefaultStepVerifierBuilder.java:801)
	at io.scalecube.services.ServiceRemoteTest.test_remote_bidi_greeting_expect_IllegalArgumentException(ServiceRemoteTest.java:286)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:513)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:170)
	at org.junit.jupiter.engine.execution.ThrowableCollector.execute(ThrowableCollector.java:40)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:166)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:113)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:58)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:113)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:121)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:121)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:121)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:121)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:55)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:43)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)
Caused by: java.lang.IllegalStateException: FluxSwitchOnFirst allows only one Subscriber
	... 71 more

**Comment by ronenhamias** [comment]
image

File: services/src/test/java/io/scalecube/services/ServiceRemoteTest.java [hunk 1]
```diff
@@ -283,7 +285,7 @@ public void test_remote_bidi_greeting_expect_IllegalArgumentException() {
     // call the service.
     StepVerifier.create(responses)
         .expectErrorMessage("IllegalArgumentException")
-        .verify(Duration.ofSeconds(3));
+        .verify(Duration.ofSeconds(300));
```
Comment (segabriel) [inline services/src/test/java/io/scalecube/services/ServiceRemoteTest.java]: is it really needed?

**Comment by ronenhamias** [comment]
It's best practice so why have our own head and tail if it comes from the reactor lib

It's just some bug in reactor lib.

File: services/src/test/java/io/scalecube/services/ServiceRemoteTest.java (hunk 1 above)
Comment (ronenhamias) [reply services/src/test/java/io/scalecube/services/ServiceRemoteTest.java]: yes we need it but lets keep it 3 and not 300
it appear to be a bug in reactor so lets just wait until its fixed.

https://github.com/reactor/reactor-core/pull/1493/files

**Comment by segabriel** [comment]
I see we use netty version `4.1.31.Final` at that time reactor-netty uses `4.1.33.Final`, we should consider how to resolve this problem.

File: services/src/test/java/io/scalecube/services/ServiceRemoteTest.java (hunk 1 above)
Comment (ronenhamias) [reply services/src/test/java/io/scalecube/services/ServiceRemoteTest.java]: https://github.com/rsocket/rsocket-java/pull/579

**Review by artem-v — APPROVED** [review]

**Pull request merged**